src/queue.rs

```diff
@@ -0,0 +1,226 @@
+                self.tail
+                    .compare_exchange(tail, node.as_ptr(), Ordering::AcqRel, Ordering::Acquire)
+                    .ok();
+
+                // Make sure the node knows it's enqueued.
+                unsafe {
+                    node.as_ref().listener.enqueue();
+                }
+
+                return node;
+            } else {
+                // The node has not been added, try again.
+                continue;
+            }
+        }
+    }
+
+    /// Pop a node from the queue.
+    pub(crate) fn pop(&self) -> Option<NonNull<Node>> {
+        loop {
+            // Get the head of the queue.
+            let head = self.head.load(Ordering::Acquire);
+
+            // If the head is null, the queue is empty.
+            if head.is_null() {
+                return None;
+            }
+
+            // Get the next pointer.
+            let next = unsafe { (*head).next.load(Ordering::Acquire) };
+
+            // Try to set the head to the next pointer.
+            if self
+                .head
+                .compare_exchange(head, next, Ordering::AcqRel, Ordering::Acquire)
+                .is_ok()
+            {
+                // If we just set the head to zero, set the tail to zero as well.
+                if next.is_null() {
+                    self.tail.store(ptr::null_mut(), Ordering::SeqCst);
+                }
+
+                // The head has been updated. Dequeue the old head.
+                let head = unsafe { NonNull::new_unchecked(head) };
+                if unsafe { head.as_ref().listener.dequeue() } {
+                    // The EventListener for this node has been dropped, free the node.
+                    unsafe { self.cache.dealloc(head) };
+                } else {
+                    // The event is still in use, return the node.
+                    return Some(head);
+                }
+            }
+        }
+    }
+
+    /// Orphan a node in the queue.
+    ///
+    /// The return value is Some if the node was notified, and true is the notification
+    /// was an additional notification.
+    pub(crate) unsafe fn orphan(&self, node: NonNull<Node>) -> Option<bool> {
+        let (needs_drop, notified) = unsafe { node.as_ref().listener.orphan() };
+
+        // If we need to deallocate the node, do so.
+        if needs_drop {
+            unsafe { self.cache.dealloc(node) };
+        }
+
+        notified
+    }
+}
+
+impl CachedNode {
+    /// Allocate a new node for a listener.
+    fn alloc(&self, listener: Listener) -> NonNull<Node> {
+        // Try to install a cached node.
+        if !self.occupied.swap(true, Ordering::Acquire) {
+            // We can now initialize the node.
+            let node_ptr = self.node.get() as *mut Node;
+
+            unsafe {
+                // Initialize the node.
+                node_ptr.write(Node::new(listener));
+
+                // Return the node.
+                NonNull::new_unchecked(node_ptr)
+            }
```

Comment:
I don't like this part at all, especially since it appears to misuse `MaybeUninit` and/or `UnsafeCell` somewhat. how about:
```rust
unsafe {
    let node_ptr: &mut MaybeUninit<_> = &mut *self.node.get();
    // Initialize the node
    node_ptr.write(Node::new(listener));
    // Return the node
    node_ptr.assume_init_mut().into()
}
```

Reply:
I've replaced the node cast to an `as_mut_ptr()` usage. Unfortunately I can't use `MaybeUninit::write()` and `MaybeUninit::assume_init_mut()` since they we introduced in Rust 1.55, which would lead to a significant MSRV bump